presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -0,0 +1,584 @@
+
+            // notify reader outside of lock
+            readWaiter.finishAndNotifyWaiters();
+
+            // notify finish listener outside of lock
+            onFinish.accept(this);
+        }
+    }
+
+    public static class LocalExchangeBufferInfo
+            implements Mergeable<LocalExchangeBufferInfo>
+    {
+        private final long bufferedBytes;
+        private final int bufferedPages;
+
+        @JsonCreator
+        public LocalExchangeBufferInfo(
+                @JsonProperty("bufferedBytes") long bufferedBytes,
+                @JsonProperty("bufferedPages") int bufferedPages)
+        {
+            this.bufferedBytes = bufferedBytes;
+            this.bufferedPages = bufferedPages;
+        }
+
+        @JsonProperty
+        public long getBufferedBytes()
+        {
+            return bufferedBytes;
+        }
+
+        @JsonProperty
+        public int getBufferedPages()
+        {
+            return bufferedPages;
+        }
+
+        @Override
+        public LocalExchangeBufferInfo mergeWith(LocalExchangeBufferInfo other)
+        {
+            return new LocalExchangeBufferInfo(bufferedBytes + other.getBufferedBytes(), bufferedPages + other.getBufferedPages());
+        }
+    }
+
+    @ThreadSafe
+    public static class LocalExchangeSinkFactory
+            implements Closeable
+    {
+        private final LocalExchange exchange;
+
+        @GuardedBy("this")
+        private boolean closed;
+
+        private LocalExchangeSinkFactory(LocalExchange exchange)
+        {
+            this.exchange = exchange;
+        }
+
+        public List<Type> getTypes()
+        {
+            return exchange.getTypes();
+        }
+
+        public LocalExchangeSink createSink()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                checkState(!closed, "Factory already closed");
+            }
+            return exchange.createSink(this);
+        }
+
+        public LocalExchangeSinkFactory duplicate()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                checkState(!closed, "Factory already closed");
+            }
+            return exchange.createSinkFactory();
+        }
+
+        @Override
+        public void close()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                if (closed) {
+                    return;
+                }
+                closed = true;
+            }
+            exchange.sinkFactoryClosed(this);
+        }
+
+        public void noMoreSinkFactories()
+        {
+            checkNotHoldsLock(this);
+
+            exchange.noMoreSinkFactories();
+        }
+    }
+
+    @ThreadSafe
+    public static class LocalExchangeSink
+    {
+        private final LocalExchange exchange;
+        private final Consumer<Page> sink;
+
+        @GuardedBy("this")
+        private boolean finished;
```

Comment:
Assuming the complexity of this class does not increase as a result of further code reviews, `AtomicBoolean` seems to be easier to use and produce cleaner code than `sync(this)` and a `boolean`.